### PR TITLE
Refactoring link modal to remove it from prosemirror plugin.

### PR DESCRIPTION
**Pull request by jpuri**
Refactoring Link modal to remove plugin.

### CODE REVIEW

File: packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.ts [hunk 1]
```diff
@@ -48,6 +48,7 @@ export function insertLinkToFile(
 }
 
 export function unmountLinkForm(view: EditorView) {
+  console.log('into unmountLinkForm')
```
Comment (ncphillips) [inline packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.ts]:
```suggestion
```

File: packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.ts [hunk 2]
```diff
@@ -170,43 +177,28 @@ export function stopEditingLink(state: EditorState, dispatch: Dispatch | null) {
...
-        const from = i
-        const to = from + node.nodeSize
-        tr.removeMark(from, to, linkMark)
-      }
-    })
-
-    dispatch(tr.scrollIntoView())
+    if (mark) {
+      dispatch(
+        state.tr.removeMark(mark.from, mark.to, mark.mark).scrollIntoView()
```
Comment (ncphillips) [inline packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.ts]: Amazing 🙌

**Comment by ncphillips** [comment]
This looks good from what I've seen. Haven't had time to test

**Comment by DirtyF** [comment]
I did run a quick test and it explodes when you open the link modal and click the delete button:

```
RangeError: Applying a mismatched transaction
EditorState.applyInner
node_modules/prosemirror-state/dist/index.es.js:851
  848 | 
  849 | // : (Transaction) → EditorState
  850 | EditorState.prototype.applyInner = function applyInner (tr) {
> 851 |   if (!tr.before.eq(this.doc)) { throw new RangeError("Applying a mismatched transaction") }
  852 |   var newInstance = new EditorState(this.config), fields = this.config.fields;
  853 |   for (var i = 0; i < fields.length; i++) {
  854 |     var field = fields[i];
View compiled
EditorState.applyTransaction
node_modules/prosemirror-state/dist/index.es.js:819
  816 | EditorState.prototype.applyTransaction = function applyTransaction (rootTr) {
  817 |   if (!this.filterTransaction(rootTr)) { return {state: this, transactions: []} }
  818 | 
> 819 |   var trs = [rootTr], newState = this.applyInner(rootTr), seen = null;
  820 |   // This loop repeatedly gives plugins a chance to respond to
  821 |   // transactions as new transactions are added, making sure to only
  822 |   // pass the transactions the plugin did not see before.
View compiled
EditorState.apply
node_modules/prosemirror-state/dist/index.es.js:795
  792 | // :: (Transaction) → EditorState
  793 | // Apply the given transaction to produce a new state.
  794 | EditorState.prototype.apply = function apply (tr) {
> 795 |   return this.applyTransaction(tr).state
  796 | };
  797 | 
  798 | // : (Transaction) → bool
View compiled
EditorView.dispatchTransaction
/Users/frank/tina/tinacms/packages/@tinacms/fields/build/index.js:1
....
```

**Comment by ncphillips** [comment]
Changed the base to `master` and rebased. I'll try to fix that build issue.

**Comment by jpuri** [comment]
Hey @DirtyF , @ncphillips : This is still WIP, I will do some more work on this today.

**Comment by jpuri** [comment]
@ncphillips , @DirtyF : PR is ready for review.
It reduces about 350 lines of code !

File: packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.test.ts [hunk 3]
```diff
@@ -18,99 +18,13 @@ limitations under the License.
...
 import { defaultSchema } from '../default-schema'
 
-const { forDoc, shouldNotRun, doc, p, link, text } = new PMTestHarness(
-  defaultSchema
-)
-
-/**
- * stopEditingLink(state, dispatch)
- */
-
```
Comment (jpuri) [inline packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.test.ts]: Fn `stopEditingLink` no longer exists.

File: packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.test.ts [hunk 4]
```diff
@@ -203,66 +92,28 @@ describe('removeLinkBeingEdited', () => {
...
               title: 'Link Title',
             })
           )
         )
       )
   })
 })
-
-/**
- * startEditingLink(state, dispatch)
```
Comment (jpuri) [inline packages/@tinacms/fields/src/Wysiwyg/commands/link-commands.test.ts]: Fn `startEditingLink` removed.

File: packages/@tinacms/fields/src/Wysiwyg/state/plugins/links/index.tsx [hunk 5]
```diff
@@ -16,34 +16,55 @@ limitations under the License.
...
-          linkForm.unmount()
-          return false
-        default:
-          return true
-      }
-    },
-    view(editorView: any) {
-      insertElBefore(renderTarget, editorView.dom)
-      linkForm = new LinkFormController(renderTarget, editorView as any, theme)
-      return new LinkView(editorView as any, schema, renderTarget)
```
Comment (jpuri) [inline packages/@tinacms/fields/src/Wysiwyg/state/plugins/links/index.tsx]: State if link form is visible is saved in link plugin in this field `show_link_toolbar`.

**Comment by ncphillips** [comment]
There's a bug when you create a link using the menu. The overlay is rendered at the top of the document, where the Menu component is.

[Here's a recording](https://www.loom.com/share/d551b0e753d54b7faeccd946c745b7ea)

**Comment by jpuri** [comment]
Hey @ncphillips : I just pushed fix for modal positioning issue metioned above.

**Comment by ncphillips** [comment]
Found a couple more issues:

1. When removing a link using the menu, the modal pops up
2. When using the arrow keys to move into a link opens the modal and we can't edit the text. Seems like the modal opens no matter how the selection hits the link.

https://www.loom.com/share/8af8d3752d394314b4c91884a1701e89

**Comment by ncphillips** [comment]
I'm going to take care of these "conflicts"

**Comment by jpuri** [comment]
> Found a couple more issues:
> 
> 1. When removing a link using the menu, the modal pops up
> 2. When using the arrow keys to move into a link opens the modal and we can't edit the text. Seems like the modal opens no matter how the selection hits the link.
> 
> https://www.loom.com/share/8af8d3752d394314b4c91884a1701e89

1. Link menu will be disabled if link is currently selected in editor. Ref: https://tinacms.slack.com/archives/CSJANKXQA/p1583857130033100
2. Link modal will open even while navigating using arrow keys but inputs will not be focused.

Hey @spbyrne : should inputs get focus when user open link modal by clicking on link ?

File: packages/react-tinacms-editor/src/state/plugins/Menu/Menu.tsx [hunk 6]
```diff
@@ -86,7 +86,11 @@ const LinkControl = markControl({
   isDisabled: (view: EditorView) => {
     const { state } = view
     const { selection, schema } = state
-    const linkPresent = selection.$anchor
+    const { anchor, head } = view.state.selection
+    let pos = anchor < head ? anchor : head
+    pos = selection.empty ? pos : pos + 1
```
Comment (ncphillips) [inline packages/react-tinacms-editor/src/state/plugins/Menu/Menu.tsx]: This is interesting. Two questions:

Do you know why the `anchor` might be before the `head` sometimes?

In what situation would the selection be empty? Is it when you click at the start of the link?

File: packages/react-tinacms-editor/src/state/plugins/links/index.tsx [hunk 7]
```diff
@@ -54,15 +55,26 @@ export function links(): Plugin {
...
+        const { anchor, head } = state.selection
+        let pos = anchor < head ? anchor : head
+        pos = state.selection.empty ? pos : pos + 1
+
         if (
-          state.selection.$anchor
+          state.doc
+            .resolve(pos)
             .marks()
             .some((mark: Mark) => mark.type === state.schema.marks.link)
```
Comment (ncphillips) [inline packages/react-tinacms-editor/src/state/plugins/links/index.tsx]: Seeing as this logic is shared between here and `Menu`, maybe it should be in a helper funciton. Maybe something like `checkForMark(mark: MarkType)`. I'm only slightly worried about removing the duplication. I'm more concerned with writing some tests explaining edge cases that it's trying to catch.

File: packages/react-tinacms-editor/src/state/plugins/Menu/Menu.tsx (hunk 6 above)
Comment (jpuri) [reply packages/react-tinacms-editor/src/state/plugins/Menu/Menu.tsx]: Hey @ncphillips : anchor < head is test is selected in reverse direction like this: 
sel

Selection is `empty` in prose-mirror when its only 1 position, its not extended over text. It is always empty in case of click.

File: packages/react-tinacms-editor/src/state/plugins/links/index.tsx (hunk 7 above)
Comment (jpuri) [reply packages/react-tinacms-editor/src/state/plugins/links/index.tsx]: Make sense, I will do this cleanup 👍

File: packages/react-tinacms-editor/src/state/plugins/Menu/Menu.tsx (hunk 6 above)
Comment (ncphillips) [reply packages/react-tinacms-editor/src/state/plugins/Menu/Menu.tsx]: Oh wicked! Thanks 🙌

File: packages/react-tinacms-editor/src/state/plugins/links/index.tsx (hunk 7 above)
Comment (jpuri) [reply packages/react-tinacms-editor/src/state/plugins/links/index.tsx]: Hey @ncphillips : PR is updated.

File: packages/react-tinacms-editor/src/state/plugins/Menu/Image/ImageMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    imageUpload().then(url => {
+      console.log(url)
+    })
+  }
+
+  return (
+    <MenuButton onClick={uploadImageFn}>
+      <MediaIcon />
+    </MenuButton>
+  )
```
Comment (ncphillips) [inline packages/react-tinacms-editor/src/state/plugins/Menu/Image/ImageMenu.tsx]: Did you mean to add the image upload stuff in this PR?

File: packages/react-tinacms-editor/src/state/plugins/links/index.tsx [hunk 9]
```diff
@@ -56,6 +81,17 @@ export function links(schema: Schema, theme?: any): Plugin {
         shiftKey = e.shiftKey
         return false
       },
+      handleClickOn(view: EditorView, _1: any) {
+        const { dispatch, state } = view
+        const { tr, schema, selection } = state
+        if (
+          !selection.$head
+            .marks()
+            .some((mark: Mark) => mark.type === schema.marks.link)
```
Comment (ncphillips) [inline packages/react-tinacms-editor/src/state/plugins/links/index.tsx]:
```suggestion
          !isMarkPresent(state, schema.marks.link)
```

File: packages/react-tinacms-editor/src/state/plugins/Menu/Image/ImageMenu.tsx (hunk 8 above)
Comment (jpuri) [reply packages/react-tinacms-editor/src/state/plugins/Menu/Image/ImageMenu.tsx]: Oh this got added as I was jumping between 2 branches. I can remove it from this PR.
